astropy/coordinates/solar_system.py

```diff
@@ -246,12 +256,37 @@ def get_body_barycentric(body, time, ephemeris=None):
             # passed in a kernel chain
             kernel_spec = body
 
-        cartesian_position_body = sum([kernel[pair].compute(jd1, jd2)
-                                       for pair in kernel_spec])
-
-        barycen_to_body_vector = u.Quantity(cartesian_position_body, unit=u.km)
-
-    return CartesianRepresentation(barycen_to_body_vector)
+        body_posvel_bary = np.zeros((2 if get_velocity else 1, 3) +
+                                    getattr(jd1, 'shape', ()))
+        for pair in kernel_spec:
+            spk = kernel[pair]
+            if spk.data_type == 3:
+                # Type 3 kernels contain both position and velocity.
+                posvel = kernel.compute(jd1, jd2)
+                if get_velocity:
+                    body_posvel_bary += posvel.reshape(2, 3, posvel.shape[1:])
+                else:
+                    body_posvel_bary[0] += posvel[:4]
+            else:
+                # spk.generate first yields the position and then the
+                # derivative. If no velocities are desired, body_posvel_bary
+                # has only one element and thus the loop ends after a single
+                # iteration, avoiding the velocity calculation.
+                for body_p_or_v, p_or_v in zip(body_posvel_bary,
+                                               spk.generate(jd1, jd2)):
```

Comment:
This is going to break for N-d arrays of times I believe. ```spk.generate``` wants 1D arrays. The type 3 code above will break for similar reasons too I suspect.

I realise of course that this means the _existing_ codebase also has this bug.